### ISSUE TITLE
MQTTnet.Exceptions.MqttProtocolViolationException

### ISSUE THREAD

**Issue by mvenditto** [chkr1011/MQTTnet#1028]
### Describe the bug
I'm getting an `MQTTnet.Exceptions.MqttProtocolViolationException' in MQTTnet.dll ("The ReasonCode must be set for MQTT version 5.")` executing the below repro code on NET Core 3.1 ConsoleApp / MQTTnet version 3.1.13.
After actually receiving the message, the client gets disconnected with `DisconnectType=NotClean`.
Am i doing something wrong or is this an actual issue?

Maybe related to #737, #1022?

Thanks in advance

### Which project is your bug related to?
- Client
- Server

#### call stack:
```batch
>	MQTTnet.dll!MQTTnet.Formatter.V5.MqttV500PacketEncoder.ThrowReasonCodeNotSetException()	
 	MQTTnet.dll!MQTTnet.Formatter.V5.MqttV500PacketEncoder.EncodePubRelPacket(MQTTnet.Packets.MqttPubRelPacket packet = {unknown}, MQTTnet.Formatter.IMqttPacketWriter packetWriter = {unknown})	
 	MQTTnet.dll!MQTTnet.Formatter.V5.MqttV500PacketEncoder.EncodePacket(MQTTnet.Packets.MqttBasePacket packet = {unknown}, MQTTnet.Formatter.IMqttPacketWriter packetWriter = {unknown})	
 	MQTTnet.dll!MQTTnet.Formatter.V5.MqttV500PacketEncoder.Encode(MQTTnet.Packets.MqttBasePacket packet = {unknown})	
 	MQTTnet.dll!MQTTnet.Formatter.V5.MqttV500PacketFormatter.Encode(MQTTnet.Packets.MqttBasePacket mqttPacket = {unknown})	
 	MQTTnet.dll!MQTTnet.Formatter.MqttPacketFormatterAdapter.Encode(MQTTnet.Packets.MqttBasePacket packet = {unknown})	
 	MQTTnet.dll!<SendPacketAsync>d__33.MoveNext()	
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start(TStateMachine stateMachine = {unknown})	
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start(TStateMachine stateMachine = {unknown})	
 	MQTTnet.dll!MQTTnet.Adapter.MqttChannelAdapter.SendPacketAsync(MQTTnet.Packets.MqttBasePacket packet = {unknown}, System.TimeSpan timeout = {unknown}, System.Threading.CancellationToken cancellationToken = {unknown})	
 	MQTTnet.dll!<SendAsync>d__45.MoveNext()	
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start(TStateMachine stateMachine = {unknown})	
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start(TStateMachine stateMachine = {unknown})	
 	MQTTnet.dll!MQTTnet.Server.MqttClientConnection.SendAsync(MQTTnet.Packets.MqttBasePacket packet = {unknown})	
 	MQTTnet.dll!<SendPendingPacketsAsync>d__44.MoveNext()	
 	System.Private.CoreLib.dll!AsyncStateMachineBox`1.ExecutionContextCallback(System.Object s = {unknown})	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext = {unknown}, System.Threading.ContextCallback callback = {unknown}, System.Object state = {unknown})	
 	System.Private.CoreLib.dll!AsyncStateMachineBox`1.MoveNext(System.Threading.Thread threadPoolThread = {unknown})	
 	System.Private.CoreLib.dll!AsyncStateMachineBox`1.MoveNext()	
 	System.Private.CoreLib.dll!<>c.AnonymousMethod(System.Action innerContinuation = {unknown}, System.Threading.Tasks.Task innerTask = {unknown})	
 	System.Private.CoreLib.dll!ContinuationWrapper.Invoke()	
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action = {unknown}, bool allowInlining = {unknown})	
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(System.Object continuationObject = {unknown})	
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.FinishContinuations()	
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task`1.TrySetResult(TResult result = {unknown})	
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result = {unknown})	
 	MQTTnet.dll!<WaitOneAsync>d__4.MoveNext()	
 	System.Private.CoreLib.dll!AsyncStateMachineBox`1.ExecutionContextCallback(System.Object s = {unknown})	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext = {unknown}, System.Threading.ContextCallback callback = {unknown}, System.Object state = {unknown})	
 	System.Private.CoreLib.dll!AsyncStateMachineBox`1.MoveNext(System.Threading.Thread threadPoolThread = {unknown})	
 	System.Private.CoreLib.dll!AsyncStateMachineBox`1.MoveNext()	
 	System.Private.CoreLib.dll!<>c.AnonymousMethod(System.Action innerContinuation = {unknown}, System.Threading.Tasks.Task innerTask = {unknown})	
 	System.Private.CoreLib.dll!ContinuationWrapper.Invoke()	
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.System.Threading.IThreadPoolWorkItem.Execute()	
 	System.Private.CoreLib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	
 	System.Private.CoreLib.dll!System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()	
```
### Code example
```csharp
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Client.Options;
using MQTTnet.Formatter;
using MQTTnet.Packets;
using MQTTnet.Protocol;
using MQTTnet.Server;

namespace MQTTNetIssueRepro
{
    class Program
    {
        static async Task Main(string[] _)
        {
            var factory = new MqttFactory();
            
            var serverOptions = new MqttServerOptionsBuilder()
                .WithDefaultEndpointPort(1883)
                .WithClientId("TestServer")
                .WithDefaultEndpointReuseAddress()
                .WithPersistentSessions()
                .Build();

            var clientOptions = new MqttClientOptionsBuilder()
                .WithProtocolVersion(MqttProtocolVersion.V500)
                .WithClientId("TestClient")
                .WithTcpServer("localhost", 1883)
                .WithCleanSession(false)
                .WithSessionExpiryInterval(0xFFFFFFFF)
                .Build();

            var server = factory.CreateMqttServer();
            server.ClientDisconnectedHandler = new MqttServerClientDisconnectedHandlerDelegate(args =>
            {
                Console.WriteLine($"client {args.ClientId} disconnected ({args.DisconnectType})");
            });
            await server.StartAsync(serverOptions);

            var client = factory.CreateMqttClient();

            client.UseApplicationMessageReceivedHandler(msg =>
            {
                Console.WriteLine($"Received msg, payload len: " + msg?.ApplicationMessage?.Payload?.Length);
            });

            await client.ConnectAsync(clientOptions, CancellationToken.None);
            
            await client.SubscribeAsync(
                new MqttTopicFilter
                {
                    Topic = "/test/+",
                    QualityOfServiceLevel = MqttQualityOfServiceLevel.ExactlyOnce
                }
            );

            var message = new MqttApplicationMessageBuilder()
                .WithTopic("/test/test2")
                .WithPayload("{ \"test\": 42 }")
                .WithPayloadFormatIndicator(MqttPayloadFormatIndicator.CharacterData)
                .WithContentType("application/json")
                .WithExactlyOnceQoS()
                .Build();

            message.UserProperties = new List<MqttUserProperty>
            {
                new MqttUserProperty("TestProp", "TestPropValue")
            };

            await server.PublishAsync(message);
            
            Console.WriteLine("done");
            Console.ReadLine();
            client.Dispose();
            await server.StopAsync();
        }
    }
}
``` 
### Output
``` 
Received msg, payload len: 14
client TestClient disconnected (NotClean)
```

**Comment by chkr1011** [comment]
This will be fixed with #1022.

**Comment by SeppPenner** [comment]
@chkr1011 Can be closed, I guess? Will be fixed in the next release.
